### PR TITLE
GODRIVER-2685 Simplify the writeconcern API.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-2685](https://jira.mongodb.org/browse/GODRIVER-2685)

## Summary
Remove the functional options pattern, export all fields in the `WriteConcern` struct so users can set them directly, and add package functions for quickly specifying common write concerns `w: "majority"`, `w: 0`, and `w: 1`. Since users can now set any value to the exported `W` field, add type validation logic in the `MarshalBSONValue` function that returns an error if `W` is not a supported type, and add tests for the new error logic.

## Background & Motivation
The `writeconcern` package and types use a lot of unnecessary APIs that can confuse users. Additionally, there are no helper functions for specifying common write concerns like there are for the `readconcern` and `readpref` packages. Remove the functional options pattern, export all fields in the `WriteConcern` struct so users can set them directly, and add package functions for quickly building common write concerns `w: "majority"`, `w: 0`, and `w: 1`.

### CODE REVIEW

File: mongo/writeconcern/writeconcern.go [hunk 1]
```diff
@@ -17,34 +21,126 @@ import (
...
+	// For more information about the "w" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#w-option
+	W interface{}
+
+	// Journal requests acknowledgment from MongoDB that the write operation has been written to the
+	// on-disk journal. It sets the "j" option in a MongoDB write concern.
+	//
+	// For more information about the "j" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#j-option
+	Journal *bool
```
Comment (qingyang-hu) [inline mongo/writeconcern/writeconcern.go:65]: What are your thoughts about keeping using `J` as it is described so in the docs?

Comment (qingyang-hu) [reply mongo/writeconcern/writeconcern.go:65]: If I understand it correctly, we use a pointer here to make it omittable. However, it seems syntactically odder to me. Do you think allowing a tag or using a tri-valued type helps?

File: mongo/writeconcern/writeconcern.go [hunk 2]
```diff
@@ -17,34 +21,126 @@ import (
...
-	wTimeout time.Duration
+// Majority returns a WriteConcern that requests acknowledgment that write operations have been
+// durably committed to the calculated majority of the data-bearing voting members. If journaling is
+// enabled on mongod, nodes only acknowledge when the write operation has been written to the
+// on-disk journal.
+//
+// For more information about write concern "w: majority", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-majority-
+func Majority() *WriteConcern {
+	return &WriteConcern{W: "majority"}
```
Comment (qingyang-hu) [inline mongo/writeconcern/writeconcern.go:86]: Do we need to make `"majority"` a constant?

Comment (prestonvasquez) [inline mongo/writeconcern/writeconcern.go:86]: Should these constructors be chainable?

```go
writeconern.Majority().Journal()
```

File: mongo/writeconcern/writeconcern.go [hunk 3]
```diff
@@ -224,6 +391,8 @@ func (wc *WriteConcern) WithOptions(options ...Option) *WriteConcern {
 }
 
 // AckWrite returns true if a write concern represents an acknowledged write
+//
+// Deprecated: AckWrite will not be supported in Go Driver 2.0.
```
Comment (prestonvasquez) [inline mongo/writeconcern/writeconcern.go:395]: Should we include a migration suggestion?
```suggestion
// Deprecated: AckWrite will not be supported in Go Driver 2.0. Call "Acknowledge" method directly. 
// For example:
//
//        wc.Acknowledge()
```

File: mongo/writeconcern/writeconcern.go [hunk 4]
```diff
@@ -17,34 +21,126 @@ import (
...
-	j bool
+	// W requests acknowledgment that the write operation has propagated to a specified number of
+	// mongod instances or to mongod instances with specified tags. It sets the the "w" option
+	// in a MongoDB write concern.
+	//
+	// W must be a type string or int value.
+	//
+	// For more information about the "w" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#w-option
+	W interface{}
```
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:58]: Yes we can! I created [GODRIVER-2812](https://jira.mongodb.org/browse/GODRIVER-2812) to track that change.

File: mongo/writeconcern/writeconcern.go [hunk 5]
```diff
@@ -17,34 +21,126 @@ import (
...
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-custom-write-concern-name-
+func Custom(tag string) *WriteConcern {
+	return &WriteConcern{W: tag}
+}
+
+// W0 returns a WriteConcern that requests no acknowledgment of the write operation.
+//
+// For more information about write concern "w: 0", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-number-
+func W0() *WriteConcern {
```
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:104]: The main benefit is that `{"w": "majority"}`, `{"w": 0}`, and `{"w": 1}` are the most commonly used write concerns, so we want to provide users a quick way to define those.

Defining a convenience function like
```go
func W(int) *WriteConcern
```
that lets you specify any numeric `w` like
```go
writeconcern.W(1)
```
could also make sense, but there is already a package-level `W` function, so that's not possible currently.

Another alternative is to provide convenience functions that describe what behavior the write concern causes, like
```go
func Unacknowledged() *WriteConcern // Returns {"w": 0}
func Acknowledged() *WriteConcern   // Returns {"w": 1}
```
However, we can't extend that pattern to `Majority` because that name is also already taken.

TLDR: I picked `W0` and `W1` partly because they're common write concerns and partly to avoid naming collisions.

File: mongo/writeconcern/writeconcern.go (hunk 2 above)
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:86]: I think we should avoid adding the "chainable setters" or "builder" pattern where it's not needed because it introduces more code we have to support and makes API surface areas larger and more difficult to understand. A more idiomatic pattern in Go is to use struct literals to configure things, adding "convenience" functions for defining common configurations.

E.g. of a struct literal definition:
```go
&writeconcern.WriteConcern{
	W:       1,
	Journal: true,
}
```
E.g. of a convenience function:
```go
writeconcern.Majority()
```

In this case, the `Journal` field needs to be able to represent `true`, `false`, or unset. Using a `*bool` accomplishes that, but is syntactically awkward because Go doesn't currently allow creating pointers to constants or literals. In other places in the Go driver API, we've glossed over that issue because the package already uses the "chainable setters" pattern. However, adding the chainable setters pattern doesn't seem justified only to deal with awkward `*bool` syntax.

P.S. Check out an ongoing discussion about pointers to constants or literals [here](https://github.com/golang/go/issues/45624). A recommendation made in that thread is to define a generic function like
```go
func ptr[T any](v T) *T { return &v }
```
that would let you more easily get a pointer to a constant or literal value like
```go
&writeconcern.WriteConcern{
	W:       1,
	Journal: ptr(true),
}
```

File: mongo/writeconcern/writeconcern.go [hunk 6]
```diff
@@ -17,34 +21,126 @@ import (
...
-	// undefined behavior.
-	wTimeout time.Duration
+// Majority returns a WriteConcern that requests acknowledgment that write operations have been
+// durably committed to the calculated majority of the data-bearing voting members. If journaling is
+// enabled on mongod, nodes only acknowledge when the write operation has been written to the
+// on-disk journal.
+//
+// For more information about write concern "w: majority", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-majority-
+func Majority() *WriteConcern {
```
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:85]: The "convenience" API as currently implemented in this PR is:
```go
writeconcern.W0()
writeconcern.W1()
writeconcern.Majority()
```

A reasonable alternative could be:
```go
writeconcern.Unacknowledged()
writeconcern.Acknowledged()
writeconcern.Journaled()
writeconcern.Majority()
```

Ignoring name collisions, a reasonable alternative could be:
```go
writeconcern.W(0)
writeconcern.W(1)
writeconcern.WMajority()
```

We could also consider not providing any convenience functions at all:
```go
func ptr[T any](v T) *T { return &v }
&writeconcern.WriteConcern{W: 0}
&writeconcern.WriteConcern{W: 1}
&writeconcern.WriteConcern{W: "majority"}
```

If the

File: mongo/writeconcern/writeconcern.go (hunk 3 above)
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:395]: Good idea. I will actually just replace the text with an alternative recommendation to match the format in other deprecation notices.

File: mongo/writeconcern/writeconcern.go (hunk 2 above)
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:86]: We can add it as a constant to remove the 2 places it's repeated in `writeconcern.go`. I don't think we should export that constant because we generally want people to use the `Majority` convenience function rather than the struct literal.

File: mongo/writeconcern/writeconcern.go (hunk 1 above)
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:65]: After doing some more digging, I found an interesting [Go issue](https://github.com/golang/go/issues/45624) discussing the difficulty of declaring pointers to constants or literals. That issue is a proposal to add a syntax for declaring a pointer to a constant, like `new(true)`, but also includes some suggestions about how you can define your own helpers to do the same thing. For example, if you're using Go 1.18+, you can create a `ptr` helper:
```go
func ptr[T any](v T) *T { return &v }
```
that would allow you to write `ptr(true)`.

I'm going to update this PR with some testable examples showing how users should use the API, including how to use the struct literal syntax.

File: mongo/writeconcern/writeconcern.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/writeconcern/writeconcern.go:85]: (1) and (2) are easy to understand and follow the documentation conventions. It seems reasonable to me to keep the existing API. Perhaps once W() is gone we can also include a W() constructor if there is demand.

File: mongo/writeconcern/writeconcern.go (hunk 2 above)
Comment (prestonvasquez) [reply mongo/writeconcern/writeconcern.go:86]: Thank you for the explanation, I like the pointer solution.

File: mongo/writeconcern/writeconcern.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:119]: I decided to mix (1) and (2) and ended up with:
```go
writeconcern.Unacknowledged()
writeconcern.W1()
writeconcern.Journaled()
writeconcern.Majority()
```
That gives us the API "space" to add additional convenience functions if we need to in the future.

File: mongo/writeconcern/writeconcern.go [hunk 7]
```diff
@@ -194,21 +342,40 @@ func (wc *WriteConcern) IsValid() bool {
...
+	return wc.W
 }
 
 // GetJ returns the write concern journaling level.
+//
+// Deprecated: Use the WriteConcern.J field instead.
 func (wc *WriteConcern) GetJ() bool {
-	return wc.j
+	// Treat a nil J as false. That maintains backward compatibility with the existing behavior of
+	// GetJ where unset is false. If users want the real value of J, they can access the J field.
```
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:362]: Good catch, will update the documentation and comments.

**Review by qingyang-hu — APPROVED** [review]

**Review by prestonvasquez — APPROVED** [review]
LGTM

File: mongo/writeconcern/writeconcern.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply mongo/writeconcern/writeconcern.go:125]: Looking at Java's `WriteConcern` class, [ACKNOWLEDGED](https://github.com/mongodb/mongo-java-driver/blob/0efc78c41c63e91da23532695fc31ee1dbd157d0/driver-core/src/main/com/mongodb/WriteConcern.java#L95) is actually the empty (i.e. default) write concern, and there is a separate [W1](https://github.com/mongodb/mongo-java-driver/blob/0efc78c41c63e91da23532695fc31ee1dbd157d0/driver-core/src/main/com/mongodb/WriteConcern.java#L103). While it is correct to say that the default write concern and write concern `{w: 1}` are acknowledged, I was worried using `Acknowledged` for a different acknowledged write concern than the Java driver might be confusing (`Acknowledged` is a bit confusing in general because most write concerns are acknowledged). Instead, I decided to start with the most obvious write concerns that are also consistent with the Java driver naming. That leaves the option open to define more write concerns later if we determine it's necessary.

[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/writeconcern/writeconcern.go:125]: Thanks for the explanation
